hpx/runtime/parcelset/parcel.hpp

```diff
@@ -149,182 +104,109 @@ namespace hpx { namespace parcelset
-          , typename...Args
-          , typename Enable =
-                typename std::enable_if<traits::is_action<Action>::value>::type
-          >
-        parcel(
-            std::vector<naming::id_type> const& dests,
-            std::vector<naming::address> const& addrs,
-            Action,
-            Args &&... args
-        )
-          : dests_(dests),
-            addrs_(addrs),
-            action_(new actions::transfer_action<Action>(std::forward<Args>(args)...))
-        {
-            HPX_ASSERT(is_valid());
-        }
-#endif
-
-        template <
-            typename Continuation
-          , typename Action
-          , typename...Args
-          , typename Enable
-                = typename std::enable_if<
-                    traits::is_action<Action>::value
-                 && traits::is_continuation<Continuation>::value
-                 && !std::is_same<
-                        Continuation
-                      , std::unique_ptr<actions::continuation>
-                    >::value
-                >::type
-          >
-        parcel(
-            naming::id_type const& dests,
-            naming::address const& addrs,
-            Continuation && cont,
-            Action,
-            Args &&... args
-        )
-#if defined(HPX_SUPPORT_MULTIPLE_PARCEL_DESTINATIONS)
-          : dests_(1, dests),
-            addrs_(1, addrs),
-#else
-          : dests_(dests),
-            addrs_(addrs),
-#endif
-            cont_(
-                new typename util::decay<Continuation>
-                    ::type(std::forward<Continuation>(cont))
-            ),
-            action_(new actions::transfer_action<Action>(std::forward<Args>(args)...))
-        {
-            HPX_ASSERT(is_valid());
-        }
-
-        template <
-            typename Action
-          , typename...Args
-          , typename Enable
-                = typename std::enable_if<
-                    traits::is_action<Action>::value
-                >::type
-          >
+    private:
         parcel(
-            naming::id_type const& dests,
-            naming::address const& addrs,
+            naming::gid_type&& dest,
+            naming::address&& addr,
             std::unique_ptr<actions::continuation> cont,
-            Action,
-            Args &&... args
+            std::unique_ptr<actions::base_action> act,
+            bool action_args_bitwise = false
         )
-#if defined(HPX_SUPPORT_MULTIPLE_PARCEL_DESTINATIONS)
-          : dests_(1, dests),
-            addrs_(1, addrs),
-#else
-          : dests_(dests),
-            addrs_(addrs),
-#endif
+          : data_(std::move(dest), std::move(addr)),
             cont_(std::move(cont)),
-            action_(new actions::transfer_action<Action>(std::forward<Args>(args)...))
+            action_(std::move(act)),
+            size_(0),
+            action_args_bitwise_(action_args_bitwise)
         {
-            HPX_ASSERT(is_valid());
+//             HPX_ASSERT(is_valid());
```

Comment:
You don't have to like it ... the rationale behind it was that the destination gid is potentially split before setting it. The two choices I saw here was: 1) keep the arguments around until the destination was split properly 2) construct the parcel "partially" and set the destination once it was split.
I implemented option 1) first, this lead to a lot of duplicated code when waiting (or not) for the destination to be split, that's why I chose the second option.